File: packages/contracts/python/science_workbench_contracts/protocols/fixture.py
from typing import Final, Self

from pydantic import model_validator
from pydantic_core import PydanticCustomError

from .models import (
    ActionPlan,
    ApprovalRecord,
    ExecutionRecord,
    Message,
    ProtocolModel,
    RunRecord,
    ToolGrant,
)
from .runtime import (
    RuntimeActionProposal,
    RuntimeContinue,
    RuntimeEvent,
    RuntimeTextDelta,
)
from .sse import RunEventWindow

FIXTURE_CORRELATION_ERROR: Final = "protocol_fixture_correlation"
FIXTURE_CORRELATION_MESSAGE: Final = "protocol fixture references do not correlate"


class ProtocolFixture(ProtocolModel):
    run: RunRecord
    messages: tuple[Message, ...]
    event_window: RunEventWindow
    action_plan: ActionPlan
    tool_grant: ToolGrant
    approval: ApprovalRecord
    execution: ExecutionRecord
    runtime_events: tuple[RuntimeEvent, ...]
    runtime_action_proposal: RuntimeActionProposal
    runtime_continuation: RuntimeContinue

    @model_validator(mode="after")
    def correlated_records(self) -> Self:
        run = self.run
        plan = self.action_plan
        binding = self.approval.binding
        message_ids = {message.id for message in self.messages}
        messages_match = all(
            message.org_id == run.org_id and message.run_id == run.id
            for message in self.messages
        )
        events_match = all(event.run_id == run.id for event in self.event_window.events)
        runtime_events_match = all(
            event.run_id == run.id for event in self.runtime_events
        )
        runtime_messages_match = all(
            event.message_id in message_ids
            for event in self.runtime_events
            if isinstance(event, RuntimeTextDelta | RuntimeActionProposal)
        )
        plan_matches = (
            plan.org_id == run.org_id
            and plan.run_id == run.id
            and plan.requester_id == run.requester_id
            and self.tool_grant.org_id == plan.org_id
            and self.tool_grant.project_id == plan.project_id
        )
        binding_matches = (
            binding.org_id == plan.org_id
            and binding.project_id == plan.project_id
            and binding.run_id == plan.run_id
            and binding.requester_id == plan.requester_id
            and binding.action_plan_id == plan.id
            and binding.plan_digest == plan.plan_digest
        )
        execution_matches = (
            self.execution.org_id == run.org_id
            and self.execution.run_id == run.id
            and self.execution.action_plan_id == plan.id
        )
        proposal = self.runtime_action_proposal
        continuation = self.runtime_continuation
        runtime_matches = (
            proposal in self.runtime_events
            and proposal.run_id == run.id
            and proposal.message_id in message_ids
            and continuation.org_id == run.org_id
            and continuation.run_id == run.id
            and continuation.message_id == proposal.message_id
            and continuation.action_id == proposal.action_id
            and continuation.execution_id == self.execution.id
            and continuation.result.execution_id == self.execution.id
        )
        if not all(
            (
                messages_match,
                events_match,
                runtime_events_match,
                runtime_messages_match,
                plan_matches,
                binding_matches,
                execution_matches,
                runtime_matches,
            )
        ):
            raise PydanticCustomError(
                FIXTURE_CORRELATION_ERROR, FIXTURE_CORRELATION_MESSAGE
            )
        return self

File: packages/contracts/src/protocols/fixture.ts
import { z } from "zod"

import {
  ActionPlanSchema,
  ExecutionSchema,
  MessageSchema,
  ProtocolApprovalSchema,
  ProtocolRunSchema,
  ToolGrantSchema,
} from "./models"
import { RuntimeActionProposalSchema, RuntimeContinueSchema, RuntimeEventSchema } from "./runtime"
import { RunEventWindowSchema } from "./sse"

export const ProtocolFixtureSchema = z
  .strictObject({
    run: ProtocolRunSchema,
    messages: z.array(MessageSchema).readonly(),
    event_window: RunEventWindowSchema,
    action_plan: ActionPlanSchema,
    tool_grant: ToolGrantSchema,
    approval: ProtocolApprovalSchema,
    execution: ExecutionSchema,
    runtime_events: z.array(RuntimeEventSchema).readonly(),
    runtime_action_proposal: RuntimeActionProposalSchema,
    runtime_continuation: RuntimeContinueSchema,
  })
  .superRefine((fixture, context) => {
    const { run, action_plan: plan, approval, execution } = fixture
    const binding = approval.binding
    const proposal = fixture.runtime_action_proposal
    const continuation = fixture.runtime_continuation
    const messageIds = new Set(fixture.messages.map((message) => message.id))
    const recordsCorrelate = [
      fixture.messages.every(
        (message) => message.org_id === run.org_id && message.run_id === run.id,
      ),
      fixture.event_window.run_id === run.id,
      fixture.event_window.events.every((event) => event.run_id === run.id),
      fixture.runtime_events.every((event) => event.run_id === run.id),
      fixture.runtime_events.every(
        (event) =>
          !(event.kind === "text_delta" || event.kind === "action_proposal") ||
          messageIds.has(event.message_id),
      ),
      fixture.runtime_events.some((event) => JSON.stringify(event) === JSON.stringify(proposal)),
      plan.org_id === run.org_id,
      plan.run_id === run.id,
      plan.requester_id === run.requester_id,
      fixture.tool_grant.org_id === plan.org_id,
      fixture.tool_grant.project_id === plan.project_id,
      binding.org_id === plan.org_id,
      binding.project_id === plan.project_id,
      binding.run_id === plan.run_id,
      binding.requester_id === plan.requester_id,
      binding.action_plan_id === plan.id,
      binding.research_intent_sha256 === plan.research_intent_sha256,
      binding.plan_digest === plan.plan_digest,
      execution.org_id === run.org_id,
      execution.run_id === run.id,
      execution.action_plan_id === plan.id,
      proposal.run_id === run.id,
      messageIds.has(proposal.message_id),
      continuation.org_id === run.org_id,
      continuation.run_id === run.id,
      continuation.message_id === proposal.message_id,
      continuation.action_id === proposal.action_id,
      continuation.execution_id === execution.id,
      continuation.result.execution_id === execution.id,
    ].every(Boolean)
    if (!recordsCorrelate) {
      context.addIssue({ code: "custom", message: "protocol fixture references do not correlate" })
    }
  })
  .readonly()

export type ProtocolFixture = z.infer<typeof ProtocolFixtureSchema>
